tensorflow/compiler/xla/service/gpu/conv_algorithm_picker.cc

```diff
@@ -1036,42 +1077,21 @@ StatusOr<bool> GpuConvAlgorithmPicker::RunOnInstruction(HloInstruction* instr) {
           << " of scratch memory: " << instr->ToString()
           << " tensor_ops_enabled: " << best_algo.conv().tensor_ops_enabled();
 
-  // Replace instr with a new CustomCall which has the correct algorithm, and
-  // whose output shape has the appropriate amount of scratch memory.
+  // Set the algorithm and update the shape of the convolution Custom Call to
+  // account for the appropriate amount of scratch memory.
```

Comment:
The original code still was careful to package the convolution output in a tuple where the second element was an empty tensor (see line 1072 of the original code). I assume this is because it's possible an instruction uses the empty tensor. You should keep that code here, even though you now update the custom call in place instead of creating a new one.

Reply:
The last element in the tuple is the workspace. The current and the original implementation preserve the shapes of the tuple elements, which should make them compatible.

Reply:
Sorry, I forgot to comment before, but the current implementation doesn't seem to preserve the shapes of the last element in the tuple. Before, the last tuple element was empty but with this PR, the last tuple element in the size of the scratch space.

Reply:
Isn't that the objective though, based on the comment in line 1039? Note that in the previous implementation, the second element of the tuple returned by new_call is not accessed.

Reply:
The old code would replace

```
output_and_empty_array = (f16[1,1,1,1], u8[0]) custom-call(...)
conv_result = f16[1,1,1,1] get-tuple-element(output_and_empty_array), index=0
```

with

```
new_call = (f16[1,1,1,1], u8[123]) custom-call(...)
new_conv_result = f16[1,1,1,1] get-tuple-element(new_call), index=0
empty_constant = u8[0] constant({})
output_and_empty_array = (f16[1,1,1,1], u8[0]) tuple(new_conv_result, empty_constant)
conv_result = f16[1,1,1,1] get-tuple-element(output_and_empty_array), index=0
```

Before this PR, `output_and_empty_array` has the same shape both before and after this pass, which is `(f16[1,1,1,1], u8[0])`. Even though the custom-call itself now has a non-empty second tuple element, the pass carefully packages the output in a new tuple such that the second element is empty. 

With this PR, you should make sure this is still done. The PR currently does not guarantee that the second element of new_call is not accessed. Some previous pass may have introduced a usage of the second element, so you should ensure an empty second element still exists.

Reply:
new_call is introduced as an operand of new_tuple. Since instr (the original Custom Call) is replaced by the tuple (i.e. the user of the first elements of new_call), I don't think that new_call's final element can have an existing user. The [autotuning](https://github.com/tensorflow/tensorflow/blob/fd9d01e4c32fc8a3f3277f24b06b5360ba462a41/tensorflow/compiler/xla/service/gpu/gpu_compiler.cc#L1045) passes are followed by a pass of the [TupleSimplifier](https://github.com/tensorflow/tensorflow/blob/fd9d01e4c32fc8a3f3277f24b06b5360ba462a41/tensorflow/compiler/xla/service/gpu/gpu_compiler.cc#L1054) which I think will eliminate the dead element describing the workspace size.